File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.cpp
class Solution {
public:
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        // edge case
        if(source == destination) return true;
        // make adj list;
        vector<int> adj[n];
        for(int i = 0; i<edges.size(); i++){
            adj[edges[i][0]].push_back(edges[i][1]);
            adj[edges[i][1]].push_back(edges[i][0]);
        }
        vector<int> pred(n,-1);
        vector<bool> visited(n,false);
        queue<int> q;
        q.push(source);
        visited[source] = true;
        // bfs
        while(!q.empty()){
            int curr_vertex = q.front(); q.pop();
            for(auto adj_vertex : adj[curr_vertex]){
                if(!visited[adj_vertex]){
                    visited[adj_vertex] = true;
                    q.push(adj_vertex);
                    pred[adj_vertex] = curr_vertex;
                }
            }
        }
        return pred[destination] != -1;
    }
};

File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @param {number} source
 * @param {number} destination
 * @return {boolean}
 */
var validPath = function(n, edges, source, destination) {
    // edge case
    if(source == destination) return true;
    const visited = new Array(n).fill(false);
    const pred = new Array(n).fill(-1);
    // adj lists
    const adj = Array.from(new Array(n), ()=>{return []});
    for(let i of edges){
        adj[i[0]].push(i[1]);
        adj[i[1]].push(i[0]);
    }
    const dfs = function (v) {
        visited[v] = true;
        for(let i of adj[v]){
            if(!visited[i]){
                pred[i] = v;
                dfs(i);
            }
        }
    }
    dfs(source);
    return pred[destination] !== -1;
};